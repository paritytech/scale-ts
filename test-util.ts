/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
