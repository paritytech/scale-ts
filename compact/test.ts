import * as $ from "../mod.ts";
import { testCodec } from "../test-util.ts";

testCodec($.compactU32, [0, 1, 6, 8, 14, 16, 30, 32].map((x) => (2 ** x) - 1));
testCodec($.compactU256, [0, 6, 14, 30, 40, 64, 128, 128, 256, 536].map((x) => (1n << BigInt(x)) - 1n));
