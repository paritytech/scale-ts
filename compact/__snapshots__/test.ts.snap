export const snapshot = {};

snapshot[`\$.compactNumber 0 1`] = `00`;

snapshot[`\$.compactNumber 1 1`] = `04`;

snapshot[`\$.compactNumber 63 1`] = `fc`;

snapshot[`\$.compactNumber 255 1`] = `
fd
03
`;

snapshot[`\$.compactNumber 16383 1`] = `
fd
ff
`;

snapshot[`\$.compactNumber 65535 1`] = `
fe
ff
03
00
`;

snapshot[`\$.compactNumber 1073741823 1`] = `
fe
ff
ff
ff
`;

snapshot[`\$.compactNumber 4294967295 1`] = `
03
ff
ff
ff
ff
`;

snapshot[`\$.compactBigInt 0n 1`] = `00`;

snapshot[`\$.compactBigInt 63n 1`] = `fc`;

snapshot[`\$.compactBigInt 16383n 1`] = `
fd
ff
`;

snapshot[`\$.compactBigInt 1073741823n 1`] = `
fe
ff
ff
ff
`;

snapshot[`\$.compactBigInt 1099511627775n 1`] = `
07
ff
ff
ff
ff
ff
`;

snapshot[`\$.compactBigInt 18446744073709551615n 1`] = `
13
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compactBigInt 340282366920938463463374607431768211455n 1`] = `
33
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compactBigInt 340282366920938463463374607431768211455n 2`] = `
33
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compactBigInt 115792089237316195423570985008687907853269984665640564039457584007913129639935n 1`] = `
73
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;

snapshot[`\$.compactBigInt 224945689727159819140526925384299092943484855915095831655037778630591879033574393515952034305194542857496045531676044756160413302774714984450425759043258192756735n 1`] = `
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
ff
`;
